### DISCUSSION TITLE
How can I create a popup menu?

### DISCUSSION THREAD

**Discussion by Horbin-Magician** [slint-ui/slint#3640] [Q&A] [answered]
Using: Win11、Rust、Slint

I want to create a popup menu, and I've noticed the https://github.com/slint-ui/slint/issues/38.
Therefore, I want to use muda to create a popup menu. But I don't know how to get the window handle.

Below code shows where I neet the window handle: 
```Rust
#[cfg(target_os = "windows")]
menu.show_context_menu_for_hwnd(window.hwnd() as isize, Some(position.into()));
```

I can get the window handle use below code **before**:
```Rust
pin_window.window().with_winit_window(|winit_win: &winit::window::Window| {
    let raw_window_handle = winit_win.raw_window_handle();
    if let raw_window_handle::RawWindowHandle::Win32(win32_window_handle) = raw_window_handle {
        let hwnd = win32_window_handle.hwnd as isize;
        let position = muda::PhysicalPosition { x: 100., y: 120. };
        menu.show_context_menu_for_hwnd(hwnd as isize, Some(position.into()));
    }
});
```
**But now**, `raw_window_handle()` is a private method.
There is any way can help me create a popup menu now?

**Comment by tronical** [comment]
> But now, raw_window_handle() is a private method.

It shouldn't be, as it comes from the raw window handle trait that winit implements. Could you post the error message you're seeing?

**Reply by Horbin-Magician** [reply]
```
error[E0599]: the method `raw_window_handle` exists for reference `&Window`, but its trait bounds were not satisfied
  --> src\module\screen_shotter\pin_win.rs:83:51
   |
83 |                 let raw_window_handle = winit_win.raw_window_handle();
   |                                                   ^^^^^^^^^^^^^^^^^ method cannot be called on `&Window` due to unsatisfied trait bounds
   |
  ::: D:\Interpreter\Rust\cargo\registry\src\index.crates.io-6f17d22bba15001f\winit-0.28.6\src\window.rs:44:1
   |
44 | pub struct Window {
   | -----------------
   | |
   | doesn't satisfy `winit::window::Window: HasRawWindowHandle`
   | doesn't satisfy `winit::window::Window: HasWindowHandle`
   |
   = note: the following trait bounds were not satisfied:
           `winit::window::Window: HasWindowHandle`
           which is required by `winit::window::Window: HasRawWindowHandle`
           `&winit::window::Window: HasWindowHandle`
           which is required by `&winit::window::Window: HasRawWindowHandle`
   = help: items from traits can only be used if the trait is in scope
help: the following trait is implemented but not in scope; perhaps add a `use` for it:
   |
1  + use raw_window_handle::HasRawWindowHandle;
   |

warning: unused import: `raw_window_handle::HasRawWindowHandle`
 --> src\module\screen_shotter\pin_win.rs:7:5
  |
7 | use raw_window_handle::HasRawWindowHandle;
  |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

For more information about this error, try `rustc --explain E0599`.
```

**Reply by Horbin-Magician** [reply]
This is related dependence:
```
[dependencies]
winit = "0.28.6"
i-slint-backend-selector = "1.2.0"
i-slint-backend-winit = "1.2.0"
raw-window-handle = "0.6.0"
```

**Reply by tronical** [reply]
Ah, I think that you may have to bring the `HasRawWindowHandle` trait into view. Could you try adding a `use raw_window_handle::HasRawWindowHandle;` as suggested by the compiler?

**Reply by tronical** [reply, marked as answer]
Ooh, nevermind, I see that you're using `raw-window-handle` version `0.6.0` but `winit` `0.28.6` implements only the traits in `raw-window-handle` version `0.5.2`. Does it work if you change your `Cargo.toml` to match the `raw-window-handle` version to `0.5.2`?

**Reply by Horbin-Magician** [reply]
Ooh, `HasRawWindowHandle` have been deprecated in version `0.6.0`.
I made a stupid mistake, thank you for your help.
By the way, does Slint have plans to implement a Menu feature of its own?

**Reply by tronical** [reply]
Glad you figured it out :)

Yes, we'd like to have first-class menu support. One of the things we're still discussing in the context of this is how to best define a menu hierarchy. We'll probably tackle this as part of the Android port in https://github.com/slint-ui/slint/issues/46 . After that we need to figure out how to create the menus best for each platform. I'm a big fan of muda myself, so that would be a great option for macOS and perhaps Windows, too.

**Reply by ogoffart** [reply]
> ```
> i-slint-backend-winit = "1.2.0"
> ```

Not that we don't follow semver for the internal crates.
You should use `i-slint-backend-winit = "=1.2.0"`  with the `=` to mean that you are pinning it to a specific version.

**Comment by ogoffart** [comment]
While we still need to implement a PopupMenu builtin in Slint, you should be able to do it with the help of PopupWindow.
Not perfect and trivial, but should be working.

**Reply by Horbin-Magician** [reply]
I have tried it, but there is a problem that PopupWindow can't be shown out of the window. (or was it just another mistake I made?)

**Reply by ogoffart** [reply]
That's true, right now this is the case with the winit backend. https://github.com/slint-ui/slint/issues/2375
We would need winit to implement a way to create popup window.

**Comment by Horbin-Magician** [comment]
A new problem arose :(
This is my code:

```Rust
pin_window.on_show_menu(move |mouse_x, mouse_y| {
    let pin_window = pin_window_clone.unwrap();
    pin_window.window().with_winit_window(|winit_win: &winit::window::Window| {
        let raw_window_handle = winit_win.raw_window_handle();
        if let raw_window_handle::RawWindowHandle::Win32(win32_window_handle) = raw_window_handle {
            let hwnd = win32_window_handle.hwnd as isize;
            let position = muda::LogicalPosition { x: mouse_x, y: mouse_y };
            menu.show_context_menu_for_hwnd(hwnd as isize, Some(position.into()));
        }
    });
    println!("show_menu begin");
    // need to do something for menu event
    if let Ok(event) = MenuEvent::receiver().try_recv() {
        match event.id {
            id if id == save_item.id() => {
                println!("save_item");
            },
            id if id == minimize_item.id() => {
                println!("minimize_item");
            },
            id if id == exit_item.id() => {
                println!("exit_item");
            },
            id if id == finish_item.id() => {
                println!("finish_item");
            },
            _ => {}
        }
    }
    println!("show_menu end");
});
```

I think `menu.show_context_menu_for_hwnd(hwnd as isize, Some(position.into()));` will occupied the thread of Slint.
Once the menu loses focus or is clicked, the thread will be returned  to Slint.

But when I click "save", "minimize" and  "exit" menuitem, the output will be: 

```
show_menu begin
show_menu end
show_menu begin
save_item
show_menu end
show_menu begin
minimize_item
show_menu end
```

This means that I did not capture the menu event after the first click. This may have something to do with the way Slint handles events, but I don't know how to capture the menu event in time. Can you help me? @ogoffart @tronical

**Reply by ogoffart** [reply]
I am not sure I understand the problem.

If `MenuEvent::receiver().try_recv() `  is blocking, this will block the slint thread and the winit eventloop, so no events will be processed during that time.

**Comment by Horbin-Magician** [comment]
> I am not sure I understand the problem.
> 
> If `MenuEvent::receiver().try_recv() ` is blocking, this will block the slint thread and the winit eventloop, so no events will be processed during that time.

Is there any way can help me manually handle the eventloop？

**Reply by tronical** [reply]
Afaics the receiver() returns a crossbeam channel, so try_recv() is not blocking. What you could do is either try to move that into a thread, blockingly read from it, and forward to slint via invoke_from_event_loop or (perhaps better) try to somehow turn the crossbeam channel receiver into a future that you could poll with slint’s spawn function.

**Reply by Horbin-Magician** [reply]
Ok, thanks for your help. Maybe I need to learn more about multithreading to complete an elegant implementation.
